pkg/adapter/apiserver/adapter.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	eventsclient "github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	"github.com/knative/eventing/pkg/kncloudevents"
+	"github.com/knative/pkg/logging"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerAgentName = "resource-controller-adapter"
+	updateEventType     = "dev.knative.apiserver.object.update"
+	deleteEventType     = "dev.knative.apiserver.object.delete"
+
+	envSinkURI    = "SINK_URI"
+	envAPIVersion = "APIVERSION_"
+	envKind       = "KIND_"
+	envOwnerRef   = "OWNEREF_" // format is OWNEREF_<RESOURCEINDEX>_[APIVERSION|KIND]_<OWNEREFINDEX>
+)
+
+// Add creates new ApiServerSource controllers, one per resource,
+// and adds them to the manager.
+func Add(mgr manager.Manager) error {
+	sinkURI, defined := os.LookupEnv(envSinkURI)
+	if !defined {
+		return fmt.Errorf("required environment variable not defined '%s'", envSinkURI)
+	}
+
+	eventsClient, err := kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return err
+	}
+
+	// Create one controller per watch
```

Comment:
the handler state is not the same: the `gvk` is different. Maybe by using `pkg/controller` we can use one controller.

Reply:
I was trying to say that we do not need a controller at all. You can just construct a duck type that watches for duck type that's type/objectmeta and just register handlers for each of them. So, in your main you'd construct a kubeclient (like the k8s watcher does today), then create a dynamic client:

"k8s.io/client-go/dynamic"
dynamicClient, err := dynamic.NewForConfig(cfg)

Then for each type that you need, you'd create one of these:
tif := &duck.TypedInformerFactory{
		Client:       dynamicClient,
		Type:         &<DUCK_TYPE_FOR_ALL_OBJECTS>,
		ResyncPeriod: resyncPeriod,
		StopChannel:  stopCh,
	}

for each of the watched object gvk {
	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
	informer, lister, err := psif.Get(gvr)
        
	informer.AddEventHandler(AddFunc: messageSender, UpdateFunc: messageSender, DeleteFunc: messageSender)
}

Reply:
We still need to create a controller in order to process multiple events concurrently. `processorListener` runs in one go routine only. Am I missing something?

Reply:
I think we need at least one deployment per service account. It could be something we optimize later.

Reply:
yes I agree

Reply:
Ok I understand what Ville is saying a bit better now. The issue here is there is no need for the controller. This can just be a kube client and a `dynamic.Interface` that sets up `duck.TypedInformerFactory`

Reply:
Also, don't use `processorListener`, use the informer cache.

Reply:
I'm using the informer cache directly which under the cover uses `processorListener`. I also refactored the code to use `dynamic.Interface` and `duck.TypedInformerFactory` (not pushed yet). I still think the controller is needed to process events concurrently.